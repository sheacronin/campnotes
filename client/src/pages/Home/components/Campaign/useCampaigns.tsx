import { useState, useEffect, useCallback } from 'react';

export interface ICampaign {
  id: number;
  title: string;
}

const useCampaigns = () => {
  const [campaigns, setCampaigns] = useState<ICampaign[]>([]);

  const getCampaigns = useCallback(async () => {
    try {
      const response = await fetch('http://localhost:5000/campaigns', {
        credentials: 'include',
      });
      const data = await response.json();
      setCampaigns(data);
    } catch (error) {
      console.error(error);
    }
  }, []);

  const addCampaign = async (campaign: ICampaign) => {
    try {
      const response = await fetch('http://localhost:5000/campaigns', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(campaign),
        credentials: 'include',
      });
      const data = await response.json();
      setCampaigns((prevState) => [...prevState, data]);
    } catch (error) {
      console.error(error);
    }
  };

  const editCampaign = async (id: number, campaign: ICampaign) => {
    try {
      const response = await fetch(`http://localhost:5000/campaigns/${id}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(campaign),
      });
      const updatedCampaign = await response.json();
      setCampaigns((prevState: ICampaign[]) =>
        prevState.map((oldCampaign) =>
          oldCampaign.id === id ? updatedCampaign : oldCampaign
        )
      );
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    getCampaigns();
  }, [getCampaigns]);

  return { campaigns, addCampaign, editCampaign };
};

export default useCampaigns;
