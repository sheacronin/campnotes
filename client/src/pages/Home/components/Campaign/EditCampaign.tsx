import React, { useState } from 'react';
import { ICampaign } from './useCampaigns';

interface EditCampaignProps {
  campaign?: ICampaign;
  onSubmitForm: Function;
  closeEditForm: () => void;
}

function EditCampaign({
  campaign = { id: -1, title: '' },
  onSubmitForm,
  closeEditForm,
}: EditCampaignProps) {
  const [title, setTitle] = useState(campaign.title);

  return (
    <article>
      <form
        onSubmit={(e) => {
          e.preventDefault();
          onSubmitForm(campaign.id, { title });
        }}
      >
        <div className="form-control">
          <label htmlFor="title">Title:</label>
          <input
            type="text"
            name="title"
            id="title"
            value={title}
            onChange={(e) => setTitle(e.target.value)}
          />
        </div>
        <div className="form-control">
          <button type="submit">Save Campaign</button>
        </div>
      </form>
      <button onClick={() => closeEditForm()}>Exit</button>
    </article>
  );
}

export default EditCampaign;
