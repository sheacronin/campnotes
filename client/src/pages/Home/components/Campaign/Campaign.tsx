import React, { useState } from 'react';
import { ICampaign } from './useCampaigns';
import EditCampaign from './EditCampaign';

interface CampaignProps {
  campaign: ICampaign;
  addCampaign: Function;
  editCampaign: Function;
}

function Campaign({ campaign, addCampaign, editCampaign }: CampaignProps) {
  const [isEditing, setIsEditing] = useState(false);

  return isEditing ? (
    <EditCampaign
      campaign={campaign}
      setIsEditing={setIsEditing}
      addCampaign={addCampaign}
      editCampaign={editCampaign}
    />
  ) : (
    <article className="campaign">
      <h3>{campaign.title}</h3>
      <button onClick={() => setIsEditing(true)}>Edit</button>
    </article>
  );
}

export default Campaign;
