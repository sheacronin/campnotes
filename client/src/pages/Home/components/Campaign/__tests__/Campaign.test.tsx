import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Campaign from '../Campaign';

describe('Campaign component', () => {
  const CAMPAIGN = {
    id: 1,
    title: 'Campaign',
  };
  const addCampaign = jest.fn();
  const editCampaign = jest.fn();

  it('renders correct heading', () => {
    render(
      <Campaign
        campaign={CAMPAIGN}
        addCampaign={addCampaign}
        editCampaign={editCampaign}
      />
    );
    expect(screen.getByRole('heading').textContent).toMatch(/Campaign/);
  });

  it('matches snapshot', () => {
    const { container } = render(
      <Campaign
        campaign={CAMPAIGN}
        addCampaign={addCampaign}
        editCampaign={editCampaign}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
