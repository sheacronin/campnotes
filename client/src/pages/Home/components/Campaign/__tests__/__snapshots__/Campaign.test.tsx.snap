// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Campaign component matches snapshot 1`] = `
<div>
  <article
    class="campaign"
  >
    <h3>
      Campaign
    </h3>
    <button>
      Edit
    </button>
  </article>
</div>
`;
